packages/core/src/app/pg-core/db/chain-db.ts

```diff
@@ -1,24 +1,523 @@
+      blockNumber
+    )).filter((otherLeafIndex) => {
+      return otherLeafIndex !== leafIndex
+    })
+
+    // Need the tree height to compute sibling/parent nodes.
+    const treeHeight = await this.getBlockTreeHeight(blockNumber)
+
+    // We don't want to delete any nodes that other transactions still need.
+    // Compute the siblings for *all* other stored transactions so we're not
+    // deleting anything critical.
+    let allOtherSiblingIndices: number[] = []
+    for (const otherLeafIndex of otherLeafIndices) {
+      const otherSiblingIndices = getMerkleSiblingIndices(
+        otherLeafIndex,
+        treeHeight
+      )
+      allOtherSiblingIndices = allOtherSiblingIndices.concat(
+        otherSiblingIndices
+      )
+    }
+
+    // Now figure out which sibling indices we can safely get rid of by finding
+    // anything that other nodes don't need. This is a safe operation because
+    // `removeMerkleTreeNode` will insert any parent that could've been
+    // computed by the removed node.
+    const siblingIndices = getMerkleSiblingIndices(leafIndex, treeHeight)
+    for (const siblingIndex of siblingIndices) {
+      if (!allOtherSiblingIndices.includes(siblingIndex)) {
+        await this.removeMerkleTreeNode(blockNumber, siblingIndex)
+      }
+    }
+  }
+
+  /**
+   * Creates an inclusion proof for a given transaction.
+   * @param transaction Transaction to create a proof for.
+   * @returns the inclusion proof for that transaction.
+   */
+  public async getInclusionProof(
+    transaction: Transaction
+  ): Promise<InclusionProof> {
+    const blockNumber = transaction.block
+    const leafIndex = await this.getLeafIndex(transaction)
+    const treeHeight = await this.getBlockTreeHeight(blockNumber)
+
+    // Get the list of siblings up the tree.
+    const siblingIndices: number[] = getMerkleSiblingIndices(
+      leafIndex,
+      treeHeight
+    )
+
+    // Get the nodes necessary to generate the proof.
+    const proof: string[] = []
+    for (const siblingIndex of siblingIndices) {
+      const sibling = await this.getMerkleTreeNode(blockNumber, siblingIndex)
+
+      // Don't have the sibling, can't compute the proof.
+      if (sibling === null) {
+        throw new Error('Cannot compute inclusion proof, missing sibling node.')
+      }
+
+      proof.push(sibling)
+    }
+
+    return proof
+  }
+
+  /**
+   * Adds a block hash to the database.
+   * @param blockNumber Block to add a hash for.
+   * @param blockHash Hash to add.
+   */
+  public async addBlockHash(
+    blockNumber: number,
+    blockHash: string
+  ): Promise<void> {
+    const key = KEYS.BLOCKS.encode([blockNumber])
+    const value = Buffer.from(blockHash)
+    await this.db.put(key, value)
+  }
+
+  /**
+   * Inserts the tree height for a given block.
+   * @param blockNumber Block number to add tree height for.
+   * @param treeHeight Height of the tree.
+   */
+  public async addBlockTreeHeight(
```

Comment:
We need to store the height of the tree for each block. Otherwise we wouldn't be able to determine if a given index is valid.